### DISCUSSION TITLE
Instrument module not implemented error (B1505A Keysight device)

### DISCUSSION THREAD

**Discussion by Dianebonk** [pymeasure/pymeasure#1083] [Q&A]
Hello, I am trying to do instrementation with B1505A using python, I just run these lines :

```python
from pymeasure.instruments.agilent import AgilentB1500

# explicitly define r/w terminations; set sufficiently large timeout in milliseconds or None.
b1500=AgilentB1500("GPIB0::17::INSTR", read_termination='\r\n', write_termination='\r\n', timeout=600000)
# query SMU config from instrument and initialize all SMU instances
b1500.initialize_all_smus()
# set data output format (required!)
```

An get this error : Module B1512A is not implented yet! 

![image](https://github.com/pymeasure/pymeasure/assets/98905793/1f8af1a0-19e1-4ef9-8504-2bba4f7c9f54)

I know that I have these modules : from UNT? : modules in B1505A are: B1511B,1;B1520A,0;B1514A,0;B1514A,0;**B1512A**,0;0,0;B1512A,0;0,0;B1513C,2;0,0.

I remove the line b1512A, in the Agilent B1500 code but I get the same error I can I solve it? 

Best regards

**Comment by BenediktBurger** [comment]
I took the liberty to modify your question: You have to add triple backticks for multi line code. If you add the programming language, you get syntax highlighting, too.

As the error states, nobody implemented the _B1512A_ module yet.
You have to modify the agilentB1500 code to support that module, especially the `SMUVoltageRanging` and `SMUCurrentRanging` classes need adjustments and the `query_modules` method.

If you need help, don't hesitate to ask.

**Reply by Dianebonk** [reply]
Hello @BenediktBurger thanks for answering. For me it was implemented and it's not working, because when I open Agilent B1500 code I see that they call the module B1512A (just querry to return channel and module type) . 

![image](https://github.com/pymeasure/pymeasure/assets/98905793/c6bd8dde-00a0-4a75-8c3b-ce5f2cf73b81)


How can I implement agilentB1500 code for all my modules? and how to connect it well with the existing code. I am beginner in this instrument communication with python.

Best regards

**Reply by BenediktBurger** [reply]
Oh, I see. I'll have another look tomorrow.

**Reply by BenediktBurger** [reply]
Now I understand: You added that line `'B1512A': 'HCSMU'` in your file, didn't you?

## A general comment regarding the working:

`query_modules` is used by `initialize_all_smus`, which uses the result of the query (in your case `"HCSMU"`) in order to generate SMU channels. It calls `initialize_smu`, which in turn creates an instance of the `SMU` class.

You have to modify also the `SMU` class to accept the channel (in your case "HCSMU"). The implementation already accepts this channel type, so your modification should work.

## Problem

I identified another problem: The "B1513C" module type is not recognized. You have to add that in `query_modules` as well.

**Reply by Dianebonk** [reply]
hello @BenediktBurger thanks, for answering, 

I didn't add a line in the code, it was already there, that's why I don't understand why it doesn't recognize this module.

**I'll modify the SMU class as you said and let you know if it works for me. (And add B1513C as well)**

Best regards,

**Reply by BenediktBurger** [reply]
The current code does not contain the module 1512A:
https://github.com/pymeasure/pymeasure/blob/758f0ca629c2b7ea0a4d01b5335709ab7cf189e1/pymeasure/instruments/agilent/agilentB1500.py#L109-L117

**Comment by Dianebonk** [comment]
Thanks, I appreciate your reply.
 
 I apologize for the confusion. After re-evaluation, it appears that the module was not included in the current code. 
 
 Yesterday, I made several changes to the file to adapt it to my needs and, in doing so, mistakenly assumed that I had copied and pasted the latest version, which included module 1512A. 
Thank you for bringing this to my attention.
